File: cours-05/tris.py
def bubblesort(array):
    changed = True
    while changed:
        changed = False
        for i in range(len(array) - 1):
            if array[i] > array[i+1]:
                array[i], array[i+1] = array[i+1], array[i]
                changed = True
                
def selectionSort(array):
    for i, e in enumerate(array):
        minimum = min(range(i, len(array)), key=array.__getitem__)
        array[i], array[minimum] = array[minimum], e
        
def insertionSort(array):
    for i, v in enumerate(array):
        for j in range(i - 1, -1, -1):
            if array[j] > v:
                array[j+1] = array[j]
                array[j] = v
                
def shellSort(array):
    step = 1
    while step <= len(array) // 3:
        step = step * 3 + 1
    while step:
        for i, e in enumerate(array[step:], step):
            while i >= step and array[i - step] > e:
                array[i] = array[i - step]
                i -= step
            array[i] = e
        step = (step - 1) // 3
        
def quickSort(array):
    if len(array) <= 1:
        return array

    lesser = []
    equal = []
    greater = []
    
    pivot = array[0]
    for e in array:
        if e < pivot:
            lesser.append(e)
        elif e > pivot:
            greater.append(e)
        else:
            equal.append(e)

    return quickSort(lesser) + equal + quickSort(greater)
    
def quickSortF(array):
    if not array:
        return []
    else:
        pivot = array[0]
        lesser = [x for x in array[1:] if x < pivot]
        greater = [x for x in array[1:] if x >= pivot]
        return quickSortF(lesser) + [pivot] + quickSortF(greater)


array = [15, 5, 11, 1, 23, 19]

array = quickSortF(array)

print(array)

File: cours-05/tris.java
void bubblesort(int[] array) {
    boolean changed = true;

    while(changed) {
        changed = false;

        for(int i = 0; i < array.length - 1; i++) {
            if(array[i] > array[i+1]) {
                int temp = array[i+1];
                array[i+1] = array[i];
                array[i] = temp;
                changed = true;
            }
        }
    }
}

void selectionSort(int[] array) {
    for(int i = 0 ; i < array.length - 1; i++) {
        int smallest = Integer.MAX_VALUE;
        int smallestIndex = -1;

        for(int j = i; j < array.length; j++) {
            if(array[j] < smallest) {
                smallest = array[j];
                smallestIndex = j;
            }
        }

        int temp = array[i];
        array[i] = array[smallestIndex];
        array[smallestIndex] = temp;
    }
}

void insertionSort(int[] array) {
    for(int i = 0; i < array.length; i++) {
        int v = array[i];
        int j = i - 1;
        while(j >= 0 && array[j] > v) {
            array[j + 1] = array[j];
            j -= 1;
        }
        array[j + 1] = v;
    }
}

void shellSort(int[] array) {
    int step = array.length / 2;
    while(step > 0) {
        for(int i = step; i < array.length; i++) {
            int j = i;
            int v = array[i];
            while(j >= step && array[j - step] > v) {
                array[j] = array[j - step];
                j = j - step;
            }
            array[j] = v;
        }

        if(step == 2)
            step = 1;
        else
            step *= (5.0 / 11);
    }
}

List<Integer> quicksort(List<Integer> array) {
    if(array.size() < 2)
        return array;

    int pivot = array.getFirst();

    List<Integer> lesser = new LinkedList<>();
    List<Integer> equal = new LinkedList<>();
    List<Integer> greater = new LinkedList<>();

    for(int e : array) {
        if (e < pivot) lesser.add(e);
        else if (e > pivot) greater.add(e);
        else equal.add(e);
    }

    lesser = quicksort(lesser);
    greater = quicksort(greater);

    lesser.addAll(equal);
    lesser.addAll(greater);

    return lesser;
}

void main() {
    Integer[] array = { 16, 5, 11, 1, 23, 19 };

    List<Integer> result = quicksort(Arrays.asList(array));

    for(int i : result)
        System.out.println(i);
}